### DISCUSSION TITLE
Equivalent of `npmrc` config option for Yarn Berry/Modern

### DISCUSSION THREAD

**Discussion by simonhaenisch** [renovatebot/renovate#28817] [Suggest an Idea]
### Tell us more.

Currently we can set a default registry for repos to use via the `npmrc` config option:

```js
module.exports = {
  npmrc: 'registry=<our-registry-server>',
  npmrcMerge: true,
}
```

This works for npm and Yarn 1 but not for newer Yarn versions. Can we get an equivalent field `yarnrc` that works for yarn 2+?

btw parsing registry urls from yarnrc yaml files is already supported, see #16353. I also looked at #14756 but it seems like a different problem (related to authentication).

**Comment by rarkins** [comment]
Is it because you do not configure your `.yarnrc.yml` default registry in the repo?

**Reply by simonhaenisch** [reply]
It's in case someone forgets to do that in their repo config. (so yeah you could say it's just a repo config problem 🙃)

**Reply by rarkins** [reply]
Is setting a default registry the main or only thing? Or also credentials?

Default registry can be set using packageRules and registryUrls so that Renovate itself looks up correctly.

So then the "missing" part would be Renovate adding that to a .yarnrc.yml before calling Yarn commands

**Reply by simonhaenisch** [reply]
Ok I wasn't aware of `registryUrls`. If I do

```js
packageRules: [
  {
    matchManagers: ['npm'],
    registryUrls: ['https://our-custom-registry.net']
  }
]
```

is that equivalent to setting

```js
{ npmrc: 'registry="https://our-custom-registry.net"' }
```

as in it uses that registry unless the repo configures a different one via its own `.npmrc` file?

**Reply by rarkins** [reply]
Setting packageRules.registryUrls will override anything in `.npmrc` files

Doing so should make sure Renovate looks up from the registry you specify, but it doesn't mean that Yarn itself will do so

**Reply by simonhaenisch** [reply]
Right got it, then it's not what we need because we need Yarn to put the correct URLs in the lock file.

We can set `YARN_NPM_REGISTRY_SERVER` as an env var via `customEnvVariables` for the sidecar, however that takes precedence over the `npmRegistryServer` setting in the repo's `.yarnrc.yml`, which is undesired for us, i.e. we want to set the default but repos should be able to configure sth else.

I think `yarnrc` as a new config option might make the most sense then? (maybe just a bit unclear that it would only be for yarn berry, not yarn 1 which works with `npmrc`)

**Reply by rarkins** [reply]
We actually don't like having npmrc and would like to remove it. As we grow to dozens of package managers we don't want to add config options for each.

The solution should be this:
- Renovate configuration using packageRules (to control registryUrls) and hostRules (to control credentials)
- Renovate writes relevant information out to disk before calling Yarn

Assuming you're self-hosted, have you tried adding a `~/.yarnrc.yml` to your container and seeing if it gets picked up by Yarn commands?

BTW we don't recommend using sidecar images anymore, it's better to use binarySource=install instead of binarySource=docker.

**Reply by simonhaenisch** [reply]
Thanks for all the info, appreciate it!

Creating a `yarnrc.yml` instead of setting `customEnvVariables` is what i thought about the other day but wasn't aware of an easy way to do this without having our own sidecar image. Good that you mention the `binarySource` thing, i'll look into it 🙏
